### PR TITLE
Hooks

### PR DESCRIPTION
**Pull request by garnich**
This PR include:
- Task 3: Components part 1 (branch: components1);
- Task 4: Components part 2 (branch: components2);
- Task 5: Hooks (branch: hooks);

### CODE REVIEW

File: src/components/header/header.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                        </ModalWindow>
+                    </div>
+                    <Search/>
+                </div>
+            }
+        </div>
+    )
+}
+
+export default Header
```
Comment (DenisKarlyuk) [inline src/components/header/header.js]: wrap to `memo`

File: src/components/header/header.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    const [showDetails, setShowDetails] = useState(false);
+    
+    const handleClose = () => setShow(false);
+    const handleShow = () => setShow(true);
+    const handleShowDetails = () => setShowDetails(!!id);
+    const handleCloseDetails = () => setShowDetails(false);
+    
+    const movies = useContext(MockDataContext);
+
+    const { id } = props;
```
Comment (DenisKarlyuk) [inline src/components/header/header.js:32]: Add `id` to `PropTypes`

File: src/components/header/header.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import './header.css'
+
+const Header = (props) => {
+    const [show, setShow] = useState(false);
+    const [showDetails, setShowDetails] = useState(false);
+    
+    const handleClose = () => setShow(false);
+    const handleShow = () => setShow(true);
+    const handleShowDetails = () => setShowDetails(!!id);
+    const handleCloseDetails = () => setShowDetails(false);
```
Comment (DenisKarlyuk) [inline src/components/header/header.js]: all finc should use `useCallback`

File: src/components/header/header.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    
+    const movies = useContext(MockDataContext);
+
+    const { id } = props;
+
+    useEffect(() => {
+        handleShowDetails();
+      }, [id]);
```
Comment (DenisKarlyuk) [inline src/components/header/header.js]: `handleShowDetails` is a extra function? please update state in `useEffect`

File: src/components/header/header.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    const movieDetailsData = movies.find(item => item.id === id);
```
Comment (DenisKarlyuk) [inline src/components/header/header.js]: Please use `useMemo`

File: src/components/header/index.js [hunk 6]
```diff
@@ -0,0 +1,3 @@
+import Header from './header'
```
Comment (DenisKarlyuk) [inline src/components/header/index.js]: export { default } from './Header'

File: src/components/modalWindows/addMovie/addMovie.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
+import React from 'react'
+
+import './addMovie.css'
+
+const AddMovie = (props) => {
+  
+    const handleSignIn = (event) => {
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/addMovie/addMovie.js]: useCallback

File: src/components/modalWindows/addMovie/addMovie.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
+import React from 'react'
+
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/addMovie/addMovie.js:3]: PropTypes

File: src/components/modalWindows/addMovie/addMovie.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+          </button>
+          <button type="submit" className="btn submit">
+            Submit
+          </button>
+        </div>
+      </form>
+    )
+}
+
+export default AddMovie
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/addMovie/addMovie.js]: memo

File: src/components/modalWindows/cardOptions/cardOptions.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+const CardOptions = (props) => {
+
+  const [showEdit, setShowEdit] = useState(false);
+  const [showDelete, setShowDelete] = useState(false);
+
+  const handleShowEdit = () => setShowEdit(true);
+  const handleCloseEdit = () => setShowEdit(false);
+  const handleShowDelete = () => setShowDelete(true);
+  const handleCloseDelete = () => setShowDelete(false);
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/cardOptions/cardOptions.js]: useCallback

File: src/components/modalWindows/cardOptions/cardOptions.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
+import React, { useState } from 'react';
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/cardOptions/cardOptions.js]: PropTypes

File: src/components/modalWindows/cardOptions/cardOptions.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+          id={id} 
+          handleClose={handleCloseDelete}
+          closeOptions={handleClose}
+        />
+      </ModalWindow>
+    </div>
+  )
+}
+
+export default CardOptions
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/cardOptions/cardOptions.js]: memo

File: src/components/modalWindows/deleteMovie/deleteMovie.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/deleteMovie/deleteMovie.js:1]: same as above

File: src/components/modalWindows/editMovie/editMovie.js [hunk 14]
```diff
@@ -0,0 +1,134 @@
+import React from 'react';
+
+import './editMovie.css';
+
+const EditMovie = (props) => {
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/editMovie/editMovie.js]: You can reuse `AddMovie` component

File: src/components/modalWindows/modalWindow/modalWindow.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+    </Modal>
+  )
+}
+
+ModalWindow.propTypes = {
+  handleClose: PropTypes.func.isRequired,
+  show: PropTypes.bool.isRequired,
+}
+
+export default ModalWindow
```
Comment (DenisKarlyuk) [inline src/components/modalWindows/modalWindow/modalWindow.js]: All components that have `props` should be wrapped in `memo`.

File: src/components/movieCard/movieCard.js [hunk 16]
```diff
@@ -0,0 +1,66 @@
+import React, { useState } from 'react';
```
Comment (DenisKarlyuk) [inline src/components/movieCard/movieCard.js]: useCallback, memo

File: src/components/movieCardFilter/movieCardFilter.js [hunk 17]
```diff
@@ -0,0 +1,39 @@
+import React, { Fragment } from 'react';
+import PropTypes from 'prop-types'
+
+import './movieCardFilter.css'
+
+const MovieCardFilter = (props) => {
+
+  const buttons = [
```
Comment (DenisKarlyuk) [inline src/components/movieCardFilter/movieCardFilter.js]: all functionality/variables that don't use a component props/state should be moved out of the component

File: src/components/movieCardFilter/movieCardFilter.js [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+  })
+  return <div className="btn-group">{button}</div>
+}
+
+MovieCardFilter.propTypes = {
+    filter: PropTypes.string.isRequired,
+    onFilterChange: PropTypes.func.isRequired,
+}
+
+export default MovieCardFilter
```
Comment (DenisKarlyuk) [inline src/components/movieCardFilter/movieCardFilter.js]: memo

File: src/components/movieList/movieList.js [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+  const onFilterChange = (filter) => setFilter(filter);
+  const onSorterChange = (sorter) => setSorter(sorter);
+
+  const { onIdChange } = props;
+  const movies = useContext(MockDataContext);
+
+  if (!movies.length) return null;
+
+  const filteredMovies = filter === 'all' ? movies : movies.filter( item => item.genre.toLowerCase() === filter);
```
Comment (DenisKarlyuk) [inline src/components/movieList/movieList.js]: useMemo

File: src/components/movieList/movieList.js [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+  const sortedMovies = sorter ? filteredMovies.sort((a, b) =>  Date.parse(a.date) - Date.parse(b.date) ) : filteredMovies;
```
Comment (DenisKarlyuk) [inline src/components/movieList/movieList.js]: useMemo

File: src/components/header/header.js (hunk 1 above)
Comment (garnich) [reply src/components/header/header.js]: done

File: src/components/header/header.js (hunk 2 above)
Comment (garnich) [reply src/components/header/header.js:32]: done

File: src/components/header/header.js (hunk 3 above)
Comment (garnich) [reply src/components/header/header.js]: done

File: src/components/header/header.js (hunk 4 above)
Comment (garnich) [reply src/components/header/header.js]: done

File: src/components/header/header.js (hunk 5 above)
Comment (garnich) [reply src/components/header/header.js]: done

File: src/components/header/index.js (hunk 6 above)
Comment (garnich) [reply src/components/header/index.js]: done

File: src/components/modalWindows/addMovie/addMovie.js (hunk 7 above)
Comment (garnich) [reply src/components/modalWindows/addMovie/addMovie.js]: done

File: src/components/modalWindows/addMovie/addMovie.js (hunk 8 above)
Comment (garnich) [reply src/components/modalWindows/addMovie/addMovie.js:3]: done

File: src/components/modalWindows/addMovie/addMovie.js (hunk 9 above)
Comment (garnich) [reply src/components/modalWindows/addMovie/addMovie.js]: done

File: src/components/modalWindows/cardOptions/cardOptions.js (hunk 10 above)
Comment (garnich) [reply src/components/modalWindows/cardOptions/cardOptions.js]: done

File: src/components/modalWindows/cardOptions/cardOptions.js (hunk 12 above)
Comment (garnich) [reply src/components/modalWindows/cardOptions/cardOptions.js]: done

File: src/components/modalWindows/editMovie/editMovie.js (hunk 14 above)
Comment (garnich) [reply src/components/modalWindows/editMovie/editMovie.js]: done

File: src/components/modalWindows/deleteMovie/deleteMovie.js (hunk 13 above)
Comment (garnich) [reply src/components/modalWindows/deleteMovie/deleteMovie.js:1]: done

File: src/components/modalWindows/cardOptions/cardOptions.js (hunk 11 above)
Comment (garnich) [reply src/components/modalWindows/cardOptions/cardOptions.js]: done

File: src/components/movieList/movieList.js (hunk 20 above)
Comment (garnich) [reply src/components/movieList/movieList.js]: done

File: src/components/movieList/movieList.js (hunk 19 above)
Comment (garnich) [reply src/components/movieList/movieList.js]: done

File: src/components/movieCardFilter/movieCardFilter.js (hunk 18 above)
Comment (garnich) [reply src/components/movieCardFilter/movieCardFilter.js]: done

File: src/components/movieCardFilter/movieCardFilter.js (hunk 17 above)
Comment (garnich) [reply src/components/movieCardFilter/movieCardFilter.js]: done

File: src/components/movieCard/movieCard.js (hunk 16 above)
Comment (garnich) [reply src/components/movieCard/movieCard.js]: done

File: src/components/modalWindows/modalWindow/modalWindow.js (hunk 15 above)
Comment (garnich) [reply src/components/modalWindows/modalWindow/modalWindow.js]: done

**Comment by garnich** [comment]
Hi, @DenisKarlyuk. This PR was updated according to the comments left.

**Comment by garnich** [comment]
Also added a custom hook.

**Pull request merged**